### ISSUE TITLE
Support for nohoist / multiple versions of the same package

### ISSUE THREAD

**Issue by michelalbers** [oven-sh/bun#6850]
### What is the problem this feature would solve?

I have a monorepo with multiple frontends and multiple backend services. Some frontends are at the latest version of React and some are still using an older version. Same goes for the backends - some need older versions of express and some are using the most recent one. 

With bun it seems that all dependencies are hoisted to the root and there can only be one version per package available. I would need to circumvent that problem with a ```nohoist``` option.

### What is the feature you are proposing to solve the problem?

Introduce ```nohoist``` to give developers more control over specific hoisting needs.

### What alternatives have you considered?

Using pnpm which stores the packages with a version tag so there are no clashes. But obviously I want to try bun :)

**Comment by frankcalise** [comment]
Also experiencing a need for this for a react-native project inside of a monorepo

**Comment by keverw** [comment]
I'm playing around with Ionic Framework and Capacitor. Looks like it complains if not in there directly even though they have multi-app support. For now my workaround is just to symlink it.

**Comment by coolcorexix** [comment]
desperately need this support too :(

**Comment by Scalahansolo** [comment]
To add to the list here, this is preventing Motion from using bun as our package manager in production.

**Comment by JRhodes95** [comment]
Likewise - our use-case is a design system that uses tailwind and asks consumers to scan the library's node_modules entry for class usage - similar to [what Flowbite does](https://flowbite.com/docs/getting-started/introduction/#require-via-npm).

Hoisting by default makes this marginally harder in monorepos as the relative path changes based on where it's hoisted to.

**Comment by gaboesquivel** [comment]
+ 1

**Comment by jbeckton** [comment]
+1

I'll have to switch back to PNPM for now. Looking forward to a solution with Bun.

**Comment by bayasdev** [comment]
+1

I have different React versions across my monorepo apps and it's causing Expo builds to fail as in https://github.com/expo/expo/issues/30143#issuecomment-2457674039

**Comment by Kobusvdwalt** [comment]
+1

**Comment by paschaldev** [comment]
This is needed. In a monorepo, React Native and NativeWind hoisting causes a problem.

Allow explicit no-hoist configuration for specific packages 🙏🏾

**Comment by danieljvdm** [comment]
Yeah this is a particularly annoying problem for dealing with multiple versions of React. Ideally, you could have all your apps pinned to the same React version but, as others have mentioned, in the case of Expo this becomes difficult since releases lag (there's still not stable support for React 19 on latest Expo).

**Comment by Kitenite** [comment]
+1 experiencing issue with docker + nextjs in a monorepo

**Comment by domonkospapptronic** [comment]
+1

**Comment by kaumac** [comment]
I'm surprised and kinda disappointed that a critical issue like this one hasn't been fixed in almost 2 years...

I just moved a project to a monorepo structure and Im having version conflicts because bun is hoisting the dependencies.

Unfortunately I'll have to drop bun on the entire platform because of this 🙁

**Comment by jbeckton** [comment]
> I'm surprised and kinda disappointed that a critical issue like this one hasn't been fixed in almost 2 years...
> 
> I just moved a project to a monorepo structure and Im having version conflicts because bun is hoisting the dependencies.
> 
> Unfortunately I'll have to drop bun on the entire platform because of this 🙁

Bun is still a great runtime; can you not get by with a different package manager like pnpm?

**Comment by kaumac** [comment]
> > I'm surprised and kinda disappointed that a critical issue like this one hasn't been fixed in almost 2 years...
> > I just moved a project to a monorepo structure and Im having version conflicts because bun is hoisting the dependencies.
> > Unfortunately I'll have to drop bun on the entire platform because of this 🙁
> 
> Bun is still a great runtime; can you not get by with a different package manager like pnpm?

Yeap, that's what I did. We were using Bun as the default package manager across 6 different projects. But because we moved to monorepo we had to drop it and move to pnpm.

Bun is great! I was an early adopter and was (and still kinda am) rooting for it.

But the thing is: The runtime doesn't have enough feature parity with Node yet. So if you use any large framework like Nextjs or Expo/React Native it doesn't work.

IMHO package manager is the entry point to start widespread adoption until there's feature parity with Node on the runtime. Anyone could use it as package manager, its easy for 3rd party libraries to add a Bun section to use it as package manager and help spread user adoption. But there's a lot of teams/codebases that use monorepo, and this "bug"/ticket keeps them away.

**Comment by SebastianBoehler** [comment]
+1

**Comment by bujxke** [comment]
+1

**Comment by jpozo20** [comment]
I'm working in a monorepo with an electron app and the hoisting causes the build to fail in CI. 

Although it looks like someone is working on it, I will also have an Expo/React Native project so I guess I'll switch to pnpm until it's merged and released.

**Comment by astudentinearth** [comment]
@jpozo20 I am working in a monorepo with an Electron app as well and I couldn't get pnpm working with electron-builder (inside the monorepo) either. Ended up using yarn for the whole project instead, but unfortunately it's rather slow.

I hope this feature gets integrated into bun soon, given there is a draft PR as well maybe it's not that far away.

**Comment by guesant** [comment]
https://bun.sh/blog/bun-v1.2.19#bun-install-linker-isolated

**Comment by paschaldev** [comment]
@guesant Tried that for an electron app within a monorepo and it worked for me 

`bun i --linker=isolated`

**Comment by rochdev** [comment]
> bun i --linker=isolated

This doesn't work quite in the same way that `nohoist` does. For our use case we need actual isolation, not just links.

**Comment by combab0** [comment]
+1
